### ISSUE TITLE
ValueTuples as theory data

### ISSUE THREAD

**Issue by aL3891** [xunit/xunit#1948]
Hello,
i've been playing around with xunit and its really nice, especially flexibility with theory data!

While i was doing that though i got a little tired of declaring new object arrays all the time and i thought, it would be cool if i could define theory data more succinctly maybe using the new tuple syntax. 

so instead of this

    public static IEnumerable<object[]> Data =>
        new[] {
            new object[] { 1, 2 },
            new object[] { 2, 3 },
            new object[] { 3, 4 },
        };

i'd like to do this

    public static IEnumerable<(int,int)> Data =>
        new[] {
            ( 1, 2 ),
            ( 2, 3 ),
            ( 3, 4 ),
        };

now, i understand that xunit needs to get an object array for invoking the theory, but i wrote a little helper method that i think is a fair compromise and also is compatible with the test framework:

    public static IEnumerable<object[]> ToXUnitData<T1, T2>(this IEnumerable<ValueTuple<T1, T2>> values)
    {
        foreach (var item in values)
        {
             yield return new object[] { item.Item1, item.Item2 };
        }
    }

this allows me to this:

    public static IEnumerable<object[]> Data =>
         new[] {
            ( 1, 2 ),
            ( 2, 3 ),
            ( 3, 4 ),
        }.ToXUnitData();

Once i was on this path i figured i'd make some more for diffrent number of parameters as well, i'll post the full code below in case anyone find this useful, but if this is something you'd be interested in i can make a PR, its not very complicated stuff but i think it can be usefull to make test code easier to read.

**Comment by aL3891** [comment]
Full code for the extension methods:

    public static class XUnitDataExtensions
    {
        public static IEnumerable<object[]> ToXUnitData<T1>(this IEnumerable<T1> values)
        {
            foreach (var item in values)
            {
                yield return new object[] { item };
            }
        }

        public static IEnumerable<object[]> ToXUnitData<T1, T2>(this IEnumerable<ValueTuple<T1, T2>> values)
        {
            foreach (var item in values)
            {
                yield return new object[] { item.Item1, item.Item2 };
            }
        }

        public static IEnumerable<object[]> ToXUnitData<T1, T2, T3>(this IEnumerable<ValueTuple<T1, T2, T3>> values)
        {
            foreach (var item in values)
            {
                yield return new object[] { item.Item1, item.Item2, item.Item3 };
            }
        }

        public static IEnumerable<object[]> ToXUnitData<T1, T2, T3, T4>(this IEnumerable<ValueTuple<T1, T2, T3, T4>> values)
        {
            foreach (var item in values)
            {
                yield return new object[] { item.Item1, item.Item2, item.Item3, item.Item4 };
            }
        }

        public static IEnumerable<object[]> ToXUnitData<T1, T2, T3, T4, T5>(this IEnumerable<ValueTuple<T1, T2, T3, T4, T5>> values)
        {
            foreach (var item in values)
            {
                yield return new object[] { item.Item1, item.Item2, item.Item3, item.Item4, item.Item5 };
            }
        }

        public static IEnumerable<object[]> ToXUnitData<T1, T2, T3, T4, T5, T6>(this IEnumerable<ValueTuple<T1, T2, T3, T4, T5, T6>> values)
        {
            foreach (var item in values)
            {
                yield return new object[] { item.Item1, item.Item2, item.Item3, item.Item4, item.Item5, item.Item6 };
            }
        }

        public static IEnumerable<object[]> ToXUnitData<T1, T2, T3, T4, T5, T6, T7>(this IEnumerable<ValueTuple<T1, T2, T3, T4, T5, T6, T7>> values)
        {
            foreach (var item in values)
            {
                yield return new object[] { item.Item1, item.Item2, item.Item3, item.Item4, item.Item5, item.Item6, item.Item7 };
            }
        }
    }

**Comment by aL3891** [comment]
ok, i went a little deeper,

suppose `MemberDataAttributeBase` would look something like this

    public override IEnumerable<object[]> GetData(MethodInfo testMethod)
    {
        var type = MemberType ?? testMethod.DeclaringType;
        var accessor = GetPropertyAccessor(type) ?? GetFieldAccessor(type) ?? GetMethodAccessor(type);
        if (accessor == null)
        {
            var parameterText = Parameters?.Length > 0 ? $" with parameter types: {string.Join(", ", Parameters.Select(p => p?.GetType().FullName ?? "(null)"))}" : "";
            throw new ArgumentException($"Could not find public static member (property, field, or method) named '{MemberName}' on {type.FullName}{parameterText}");
        }

        var obj = accessor();
        if (obj == null)
            return null;
        var resultType = obj.GetType();

        if (resultType.IsGenericType && resultType.GetGenericTypeDefinition() == typeof(Task<>))
        {
            obj = resultType.GetProperty("Result").GetValue(obj);
        }

        if (obj is IEnumerable dataItems)
        {
            return dataItems.Cast<object>().Select(item => ConvertDataItem(testMethod, item));
        }

        resultType = obj.GetType().GetInterfaces().FirstOrDefault(i => i.IsGenericType && i.GetGenericTypeDefinition() == typeof(IAsyncEnumerable<>));
        if (resultType != null)
        {
            var m = this.GetType().GetMethod(nameof(GetAsyncItems));
            var dataitems = (List<object>)m.MakeGenericMethod(resultType.GetGenericArguments()).Invoke(this, new[] { obj });
            return dataitems.Select(item => ConvertDataItem(testMethod, item));
        }

        throw new ArgumentException($"Property {MemberName} on {type.FullName} did not return IEnumerable");
    }

    public List<object> GetAsyncItems<T>(IAsyncEnumerable<T> asyncEnumerable)
    {
        var res = new List<object>();
        var enumerator = asyncEnumerable.GetAsyncEnumerator();
        try
        {
            while (enumerator.MoveNextAsync().GetAwaiter().GetResult())
            {
                res.Add(enumerator.Current);
            }
        }
        finally
        {
            enumerator.DisposeAsync().GetAwaiter().GetResult();
        }

        return res;
    }

and MemberDataAttribute looked like this

    protected override object[] ConvertDataItem(MethodInfo testMethod, object item)
    {
        switch (item)
        {
            case object[] apa:
                return apa;
            //case IEnumerable e:
            //    return e.Cast<object>().ToArray(); //not sure about this one
            case ITuple tuple:
                return Enumerable.Range(0, tuple.Length).Select(x => tuple[x]).ToArray();
            default:
                return new object[] { item };
        }
    }

you would be able to do these things

    public static (int, int)[] Data2 =>
        new[] {
            ( 1, 2 ),
            ( 2, 3 ),
            ( 3, 4 ),
        };
 
    public static Task<(int, int)[]> Data3 =>
        Task.FromResult(new[] {
            ( 1, 2 ),
            ( 2, 3 ),
            ( 3, 4 ),
        });
 
    public static IEnumerable<int> Data4 => new[] { 1, 2, 3 };
 
    public static async IAsyncEnumerable<(int, int)> Data5()
    {
        yield return await Task.FromResult((1, 2));
        yield return await Task.FromResult((2, 3));
        yield return await Task.FromResult((3, 4));
        yield return await Task.FromResult((4, 5));
    }

in other words, data members could return any enumerable type, as well as tasks and IAsyncIEnumerables and they the property data could be a single value or a tuple (Value or regular) as well as an object array

i can make a proper PR if you think it would be useful, it would resolve #1698 at least :)

**Comment by bradwilson** [comment]
IMO this should be part of [TheoryData](https://github.com/xunit/xunit/blob/master/src/xunit.core/TheoryData.cs).

**Comment by aL3891** [comment]
Can you provide a little more detail? as i understand from looking at TheroryData.cs, it's basically like another tuple class, you mean there should be some conversion from tuples to TheoryData? 

Also, how would async data methods and IAsyncEnumerable fit into TheoryData?

**Comment by bradwilson** [comment]
We could add `Add` methods which took tuples rather than raw individual data, which would make it easier to reuse existing tuples.

We have no support for `IAsyncEnumerable` at all in xUnit.net. Our current v3 target API level for test authors is `netstandard2.0` which means it's not likely to be supported in v3, either. We may be able to shoe-horn it in via dynamic type discovery, but I don't want to make any promises about it now (since it can't appear anywhere in type signatures, I'm skeptical that this is actually going to be possible).

**Comment by aL3891** [comment]
i see,
how would using the Add methods look from a syntax perspective though? would you still be able to write

   public static (int, int)[] Data2 =>
       new[] {
           ( 1, 2 ),
           ( 2, 3 ),
           ( 3, 4 ),
       };

   public static Task<(int, int)[]> Data3 =>
       Task.FromResult(new[] {
           ( 1, 2 ),
           ( 2, 3 ),
           ( 3, 4 ),
       });

i see your point about async enumerable, i could just put that in a separate assembly that targets 3.0, but what about regular task returning data method? being able to have async data methods wouls be very useful if you're using data from some external source

**Comment by bradwilson** [comment]
The compiler would translate this:

```
public static TheoryData<int, int> Data2 = new TheoryData<int, int> {
    (1, 2),
    (2, 3),
    (3, 4)
};
```

Effectively into this:

```
Data2 = new TheoryData<int, int>();
Data2.Add((1, 2));
Data2.Add((2, 3));
Data2.Add((3, 4));
```

We could also add another constructor to TheoryData that took an `IEnumerable<Tuple<int, int>>` for existing data sets.

**Comment by aL3891** [comment]
Ah you mean like that, so user code would return a new TheoryData. I suppose that would not work with yeild returning methods though?

Wouldn't it be easier to let the framework figure that out? A user would have to know about Theorydata to use this. It just seems more natural to just return a tuple/single type :) 

Also how would you see this working with tasks or async methods?

**Comment by zspitz** [comment]
I usually use [a set of extension methods like the ones above](https://github.com/zspitz/ANTLR4ParseTreeVisualizer/blob/fe5f475f91cd922d944563ca844b79c84967e654/Tests.Shared/Extensions.cs), except mine return an instance of `TheoryData<>` instead of `IEnumerable<object[]>`. @bradwilson Other than strong-typing, is there any difference between the two, perhaps because `IEnumerable<object[]>` could return something for later evaluation, while `TheoryData` is populated immediately?

RE: `Add` methods -- until you get around to putting them into `TheoryData`, C# will pick up extension methods with the appropriate signature as well.

**Comment by bradwilson** [comment]
Nope, TheoryData exists solely for the strong typing.

**Comment by jemarjomores** [comment]
Why does `TheoryData` dont support `AddRange`?

**Comment by bradwilson** [comment]
Support for tuples in TheoryData is now part of v3: https://github.com/xunit/xunit/commit/ebd88d7978f31f38f4ec85e3e3330db72d9f0236

Support for a richer/more convenient API than returning `IEnumerable<object[]>` will be covered by #1758.
